### PR TITLE
Volume Initialization Module

### PR DESCRIPTION
**Pull request by ketankelkar**
##### SUMMARY

Adds a new module 'zos_volume_init' which leverages the ICKDSF tool to initialize DASD volumes. Also adds supporting functional tests for module. And lastly, adds into module_utils, a structure which encapsulates ICKDSF's init command but which is designed to expand into other ICKDSF commands in the future.

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

zos_volume_init
##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

**Comment by ketankelkar** [comment]
My PR removes the following line, but it looks like it's still present in dev -- let me know if the line should be present in mine too - can easily modify (or can be taken care of during the merge process)

https://github.com/ansible-collections/ibm_zos_core/blob/f04e2c908b5af765180d8585966b0b7c693593e8/tests/sanity/ignore-2.11.txt#L62

**Review by ddimatos — CHANGES_REQUESTED** [review]
Overall its mostly just doc comments that I am happy to look over when finalized. The technical concerns are ensure we handled all mutual exclusive options, i saw one in the code, I thought there was a another I saw in doc but could not find it. 

Ensure that we cover all our corner cases with testing, you did a great job, ill have to review tests once more to ensure we are covering negative tests. 

The last point is I think we can do more with validation early on in the module rather than letting it go unchecked which then ICKDSF also does so unchecked. Maybe use the better arg parser for that? 

I would like to try this on a native system as well.

File: plugins/modules/zos_volume_init.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+options:
+  volume_address:
+    description:
+      - 3 or 4 hexadecimal digit address of the volume to initialize.
+      - This could be a unit address or device number.
+    required: true
+    type: str
+  verify_existing_volid:
+    description:
+      - Verify that the provided volume serial matches the one found on existing volume/minidisk.
```
[the remark this answers is not in the document]
Comment (ketankelkar) [reply plugins/modules/zos_volume_init.py:44]: I like the line below I had included because the parameter seems to be a boolean (when i first picked it up, i assumed bool too)

"Note - This option is NOT a boolean, please leave it blank in order to skip the verification."

Also the other params `verify_*` are bools, which in my mind adds to the ambiguity and therefore warrants the Note

File: plugins/modules/zos_volume_init.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+      - Module fails if volser does not match.
```
[the remark this answers is not in the document]
Comment (ketankelkar) [reply plugins/modules/zos_volume_init.py:45]: I like the line below I had included because the parameter seems to be a boolean (when i first picked it up, i assumed bool too)

"Note - This option is NOT a boolean, please leave it blank in order to skip the verification."

Also the other params verify_* are bools, which in my mind adds to the ambiguity and therefore warrants the Note

File: plugins/modules/zos_volume_init.py [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+      - Note - This option is NOT a boolean, please leave it blank in order to skip the verification.
+    required: false
+    type: str
+  verify_offline:
+    description:
+      - Verify that the device is offline to all other systems.
```
[the remark this answers is not in the document]
Comment (ketankelkar) [reply plugins/modules/zos_volume_init.py:51]: The following line was included because I was made aware that there are certain system settings which can override params passed to ICKDSF -- in which case, I believe there would be a mismatch in what is specified in a zos_volume_init playbook task and what happens on the system through ickdsf -- the line was my way of explaining that potential result

"Beware, defaults set on target z/OS systems may override ICKDSF parameters."

File: plugins/modules/zos_volume_init.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+      - Beware, defaults set on target z/OS systems may override ICKDSF parameters.
+    type: bool
+    required: false
+    default: true
+  volid:
+    description:
+      - Specify the volume serial number to initialize the volume with.
+      - Expects 1-6 alphanumeric, national ($, \\#, \\@) or special characters.
+      - volid's specified with less than 6 characters are left justified and padded with blank chars (X'40').
+      - Characters are not validated so check with operating system guide for valid alphanumeric characters.
```
[the remark this answers is not in the document]
Comment (ketankelkar) [reply plugins/modules/zos_volume_init.py:61]: I chose not to add validation because I did not think it wise to potentially restrict functionality -- figured it was ok to let ickdsf or zos handle problematic input in this case

File: plugins/modules/zos_volume_init.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+    default: true
+  sms_managed:
+    description:
+      - Assigned to be managed by Storage Management System (SMS).
+    type: bool
+    required: false
+    default: true
+  verify_no_data_sets_exist:
+    description:
+      - Verify if data sets other than the VTOC index data set and/or VVDS exist on the volume to be initialized.
```
[the remark this answers is not in the document]
Comment (ketankelkar) [reply plugins/modules/zos_volume_init.py:91]: added the word target - "...exists on the target volume."

See similar comment above about including the "beware" note regarding system default overriding ickdsf.

File: plugins/modules/zos_volume_init.py [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+      - Beware that z/OS system defaults can override ICKDSF parameters.
+    required: false
+    type: bool
+    default: true
+"""
+EXAMPLES = r"""
+- name: Initialize target volume with all default options. Target volume address is '1234', set volume name to 'DEMO01'.
```
[the remark this answers is not in the document]
Comment (ketankelkar) [reply plugins/modules/zos_volume_init.py:98]: are these not covered by sanity tests :(
maybe i have to generate docs and run on those generated files

File: plugins/modules/zos_volume_init.py [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+- name: Initialize 3 new DASD volumes (0901, 0902, 0903) for use on a z/OS system as 'DEMO01', 'DEMO02', 'DEMO03'
+        using Ansible loops.
+  zos_volume_init:
+    volume_address: "090{{ item }}"
+    volid: "DEMO0{{ item }}"
+  loop: "{{ range(1, 4, 1) }}"
+"""
+RETURN = r"""
+command:
+  description: INIT command issued to ICKDSF tool.
```
[the remark this answers is not in the document]
Comment (ketankelkar) [reply plugins/modules/zos_volume_init.py:140]: This is the raw command that is passed to ICKDSF command "init". Do we want to remove most/all references to the underlying ickdsf utility? I think the suggested edit is inaccurate.

File: plugins/modules/zos_volume_init.py [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+  sample:
+      [
+        " init unit(0903) noverify noverifyoffline volid(DEMO08) - ",
+        "   nods noindex",
+      ]
+msg:
+  description: Failure message returned by module.
+  returned: failure
+  type: str
+  sample: \'Index\' cannot be False for SMS managed volumes.
```
[the remark this answers is not in the document]
Comment (ketankelkar) [reply plugins/modules/zos_volume_init.py:153]: The convention im seeing is I(param_name) so changing it to italics. The purpose of the single quotes was to highlight that index is the param name.

File: plugins/modules/zos_volume_init.py [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+rc:
+  description:
+    - return code from ICKDSF init mvs command
+  type: dict
+  returned: when ICKDSF init command is issued.
```
[the remark this answers is not in the document]
Comment (ketankelkar) [reply plugins/modules/zos_volume_init.py:158]: Not returned if the module doesn't get far enough to run ickdsf. Eg if user specifies mutually exclusive options index:false and sms_managed:true, there is no `rc`  -- I didn't think it was a great idea to make up my own rcs or recreate what ickdsf would

File: plugins/modules/zos_volume_init.py [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+content:
+  description:
+    - Raw output from ICKDSF.
+  returned: when ICKDSF init command is issued.
```
[the remark this answers is not in the document]
Comment (ketankelkar) [reply plugins/modules/zos_volume_init.py:162]: Not necessarily -- there is important content returned on failure too - eg volume_address: XYZ (too short) yields:

TASK [FAIL - invalid volume_address.] ***********************************************************************************************************************************************************************
Friday 03 March 2023  11:45:39 -0800 (0:00:11.201)       0:00:57.906 ********** 
fatal: [zos_host]: FAILED! => {"changed": false, "command": [" init unit(XYZ) noverify noverifyoffline volid(KET987) - ", "  storagegroup nods "], "content": ["1ICKDSF - MVS/ESA    DEVICE SUPPORT FACILITIES 17.0                TIME: 11:45:47        03/03/23     PAGE   1", "0        ", "0 INIT UNIT(XYZ) NOVERIFY NOVERIFYOFFLINE VOLID(KET987) -", "0  STORAGEGROUP NODS", "0ICK31023I INVALID UNITADDRESS, SYSNAME OR DDNAME SPECIFIED", "-ICK30003I FUNCTION TERMINATED. CONDITION CODE IS 12", "0          11:45:47    03/03/23", "0        ", "-ICK00002I ICKDSF PROCESSING COMPLETE. MAXIMUM CONDITION CODE WAS 12"], "msg": "Non-zero return code. See 'content' for details", "rc": 12}

File: plugins/modules/zos_volume_init.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ketankelkar) [reply plugins/modules/zos_volume_init.py:153]: Yes it does: 

TASK [test vol module.] ****************************************************************************************************************************************************
Friday 03 March 2023  11:52:44 -0800 (0:00:11.835)       0:00:11.847 ********** 
fatal: [zos_host]: FAILED! => {"changed": false, "msg": "'Index' cannot be False for SMS managed volumes."}

File: plugins/module_utils/ickdsf.py [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+
+from ansible_collections.ibm.ibm_zos_core.plugins.module_utils.zos_mvs_raw import MVSCmd  # pylint: disable=import-error
+from ansible_collections.ibm.ibm_zos_core.plugins.module_utils.dd_statement import (  # pylint: disable=import-error
+    DDStatement,
+    StdoutDefinition,
+    StdinDefinition,
+)
+
+
+def convert(module, result, args):
```
[the remark this answers is not in the document]
Comment (ketankelkar) [reply plugins/module_utils/ickdsf.py:28]: LOL good catch -- i created ickdsf.py so other command implementations could have a home -- and then i got lazy with coming up with a naming convert

File: plugins/module_utils/ickdsf.py [hunk 12]
```diff
@@ -0,0 +1,140 @@
+# Copyright (c) IBM Corporation 2020
```
[the remark this answers is not in the document]
Comment (ketankelkar) [reply plugins/module_utils/ickdsf.py:1]: My understanding for copyright years was to include each year work was done. But I was working under the assumption that that applies to edits (eg file released in 2020, edited in 2022 would have (C) 2020, 2022. But for unreleased files, I was thinking it would only be a single year  that it was released ie "(C) 2023"

File: plugins/modules/zos_volume_init.py [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+  - "Ketan Kelkar (@ketankelkar)"
+
+options:
+  volume_address:
+    description:
+      - 3 or 4 hexadecimal digit address of the volume to initialize.
+      - This could be a unit address or device number.
+    required: true
+    type: str
+  verify_existing_volid:
```
[the remark this answers is not in the document]
Comment (ketankelkar) [reply plugins/modules/zos_volume_init.py:42]: I'd considered having subopts for "verify" but I didnt like that we had mixed types:
```
verify:
  volid: "DEMO12"     <-- string
  offline: yes        <-- bool
  no_data_sets: yes   <-- bool
```

and the following would break or check if current volid is "no" depending on how no was passed (no vs "no")
```
verify:
  volid: no
  offline: yes
  no_data_sets: yes
```

so to skip volid check, the user would have to exclude `volid`:
```
verify:
  offline: yes
  no_data_sets: yes
```
but to skip any others, the user could simply toggle on/off:
```
verify:
  volid: "DEMO12"
  offline: no
  no_data_sets: no
```

Overall, I do like the idea of having a "verify" or "checks" option with suboptions with all available checks -- if i am putting too much emphasis on the mixed types (as demonstrated above) i'd be happy to make this modification

File: plugins/modules/zos_volume_init.py [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+  - "Austen Stewart (@stewartad)"
+  - "Almigdad Suliman (@Almigdad-Suliman)"
+  - "Nicholas Teves (@TODO)"
+  - "Nuoya Xie (@nxie13)"
+  - "Trevor Glassey (@tkglassey)"
+  - "Tyler Edwards (@TLEdwards-Git)"
+  - "Ketan Kelkar (@ketankelkar)"
+
+options:
+  volume_address:
```
[the remark this answers is not in the document]
Comment (ketankelkar) [reply plugins/modules/zos_volume_init.py:36]: i'd like to get an ickdsf expert to make this decision -- if ickdsf waits infinitely -- i agree with the wait_time_s approach

File: plugins/modules/zos_volume_init.py [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+    description:
+      - Specify the volume serial number to initialize the volume with.
+      - Expects 1-6 alphanumeric, national ($, \\#, \\@) or special characters.
+      - volid's specified with less than 6 characters are left justified and padded with blank chars (X'40').
+      - Characters are not validated so check with operating system guide for valid alphanumeric characters.
+      - Also referred to as volser or volume serial number.
+      - Default behavior in the case this parameter is not specified in the case of an existing device initialization is the reuse the existing volume serial.
+    required: false
+    type: str
+  vtoc_tracks:
```
[the remark this answers is not in the document]
Comment (ketankelkar) [reply plugins/modules/zos_volume_init.py:66]: i like vtoc_size -- then if the time/need comes up to be able to specify it in terms of non-track units, we wont need to change the name of vtoc_tracks

File: plugins/modules/zos_volume_init.py [hunk 16]
```diff
@@ -0,0 +1,242 @@
...
+    description:
+      - The number of tracks to initialize the VTOC with.
```
[the remark this answers is not in the document]
Comment (ketankelkar) [reply plugins/modules/zos_volume_init.py:68]: i wouldn't want to hard-code a default just to keep that decision-making capacity at the ickdsf level -- i wanted something in the description about encouraging the user to let the system make this decision

File: plugins/modules/zos_volume_init.py [hunk 17]
```diff
@@ -0,0 +1,242 @@
...
+
+- name: Initialize 3 new DASD volumes (0901, 0902, 0903) for use on a z/OS system as 'DEMO01', 'DEMO02', 'DEMO03'
+        using Ansible loops.
+  zos_volume_init:
+    volume_address: "090{{ item }}"
+    volid: "DEMO0{{ item }}"
+  loop: "{{ range(1, 4, 1) }}"
+"""
+RETURN = r"""
+command:
```
[the remark this answers is not in the document]
Comment (ketankelkar) [reply plugins/modules/zos_volume_init.py:139]: zos_blockinfile uses cmd,
zos_copy uses cmd,
zos_lineinfile uses cmd,
zos_mount uses cmd,
zos_operator uses cmd,

zos_tso_command uses command/commands. 

Changing return param name in zos_volume_init to cmd

**Comment by ketankelkar** [comment]
In regards to the mutual exclusive params, I did not see any other ones. In the ickdsf docs, dsexist and nodsexist are marked as mutually exclusive but both are encompassed by a single param (verify_volume_empty). I double checked the rest and did not come across any other instances of mutually exclusive.

In regards to testing, I included cases I could think of, happy to add any missing scenarios that are identified.

My reasoning around not doing validation early in the module was to avoid unintentionally adding restrictions on the ansible side which may have been valid input to ickdsf via non-ansible. I understand the performance gain from early validation. I think another advantage of not doing early validation is increased flexibility in our module -- however perhaps that has a negative effect of potentially decreasing portability of a playbook task from a more permissive system and a lesser one. 

Not sure how to handle this going forward, i'd be happy to add early validation around:
- address - length, ensuring hex-legal
-  volid, verify_volid - length, maybe chars
- vtoc_tracks - enforce natural numbers (not sure if that's particularly helpful)

But I am still leaning towards letting the ickdsf utility handle validation

File: plugins/module_utils/ickdsf.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (ddimatos) [reply plugins/module_utils/ickdsf.py]: @ketankelkar - I was looking at it from when the file was created by the jumpstart program, I suppose 2020 might not be when it started but i don't recall when they created it. The git blame is not helpful here because it was in another repo to begin with then copied over to ours and edited last year and this year, so your understanding correct. Without a history I am going by memory, you know better and probably can adjust for this.

File: plugins/modules/zos_volume_init.py [hunk 18]
```diff
@@ -0,0 +1,242 @@
...
+    description:
+      - Verify that the device is offline to all other systems.
+      - Beware, defaults set on target z/OS systems may override ICKDSF parameters.
+    type: bool
+    required: false
+    default: true
+  volid:
+    description:
+      - Specify the volume serial number to initialize the volume with.
+      - Expects 1-6 alphanumeric, national ($, \\#, \\@) or special characters.
```
[the remark this answers is not in the document]
Comment (ddimatos) [reply plugins/modules/zos_volume_init.py:66]: @ketankelkar - wondering what we did here with regard to the documented escapes which would not be valid in the better arg parser and documenting them is inconsistent with our other modules, this quarter we are fixing it so the better arg parser will escape special charts.

File: plugins/modules/zos_volume_init.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ddimatos) [reply plugins/modules/zos_volume_init.py:68]: I read this as the only validation happening is the padding for a volid with less than 6 chars, is there more validation happening than this? If so, it might make sense to let ICKDSF propagate errors so long as they bubble up correctly. Do you have an example of this? If it is limited to just the length and padding, to me it makes seems we could fail that fast rather than the system log a system log error for something we can handle.

File: plugins/modules/zos_volume_init.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ddimatos) [reply plugins/modules/zos_volume_init.py:105]: They should be covered under the Sanity tests, Ansible sometimes override the defaults and relaxes some rules, this could one of those that allow up to 120, i don't recall offhand. If it is passing that is fine, i tend to sent my margin at 80 just to be on the safe side. 

The doc generation catches some things , like syntax errors but I don't think the PEP8 rules are enforced

File: plugins/modules/zos_volume_init.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ddimatos) [reply plugins/modules/zos_volume_init.py:147]: I was going for masking the utilities we use, there are a few ways at looking at this, our initial intention was to shield users from the complexity of z/OS, as much a we try we can't always achieve that and terms do bubble up. 

When you don't know much about ICKDSF, not sure this means anything to any one "INIT command issued to ICKDSF tool.", not even all system folks know this utility, that was the direction I was taking. 

I think it might better to describe how we achieve creating volumes the documentation section using ICKDSF where we can be a bit more verbose over this short 1-liner, or explain it further detail here so its not so cryptic to folks.

File: plugins/modules/zos_volume_init.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ddimatos) [reply plugins/modules/zos_volume_init.py:160]: @ketankelkar what i meant was , do the `\`'s come back, i can see the single quotes are escaped but your doc has `\'Index\'` and your example interprets this so that I think the sample might be `'Index'`

File: plugins/modules/zos_volume_init.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (ddimatos) [reply plugins/modules/zos_volume_init.py:165]: If it does not get far enough to run ICKDSF, then this is also not describing the  returned value "when ICKDSF init command is issued."

The [RETURN BLOCK](https://github.com/ansible/ansible/blob/5e131a96c086eda58436429a417c8e7cf256602b/docs/docsite/rst/dev_guide/developing_modules_documenting.rst#id38) has this recommendation

> When this value is returned, such as always, changed or success. This is a string and can contain any human-readable content.

Wondering if this one can be 'on success' ?

File: plugins/modules/zos_volume_init.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (ddimatos) [reply plugins/modules/zos_volume_init.py:169]: I was approaching this as on success meaning execution , but see your point. I have seen other modules where they get verbose and I posted in the prior comment the doc to return stmts. Maybe we can do something like "returned: when initialization command ICKDSF is run", or "returned: when initialization command is run"

**Review by ddimatos — APPROVED** [review]
Feel free to book some time to 1:1 these, I think it might help get to a consensus on some of these, they are all minor either way.  The most important comment being the one about the escape chars.

**Review by ddimatos — CHANGES_REQUESTED** [review]
Just noticed no changelog fragment is in the PR.

File: plugins/modules/zos_volume_init.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (ketankelkar) [reply plugins/modules/zos_volume_init.py]: per discussions with dfsms team members, it seems relatively difficult to achieve said "nonterminating wait state".

File: plugins/modules/zos_volume_init.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (ketankelkar) [reply plugins/modules/zos_volume_init.py:66]: To the 2nd point there, the escape backslashes were from me because I wanted to ensure the doc-generation wouldn't screw them up -- seems like a misunderstanding on my part. 

The following playbook task code works as expected, no escape-s needed
```
      zos_volume_init:
        address: "904"
        verify_offline: False
        volid: "D#M@$1"
```
```
bash-4.3# opercmd 'd u,,,904,1'
EC33018A   2023097  15:27:45.00             ISF031I CONSOLE OMVS0000 ACTIVATED
EC33018A   2023097  15:27:45.00            -D U,,,904,1
EC33018A   2023097  15:27:45.00             IEE457I 15.27.45 UNIT STATUS 842
                                           UNIT TYPE STATUS        VOLSER     VOLSTATE      SS
                                           0904 3390 O             D#M@$1     PRIV/RSDNT     0
```

**Comment by ketankelkar** [comment]
Ran the zos_fetch test locally and got all passing results. The new test case closely recreates the attributes of the former data set (`IMSTESTL.IMS01.SPOOL1`). See snippets below for comparisons. The content copied over to the new data set is different.

`IMSTESTL.IMS01.SPOOL1` : 
```
                              Data Set Information                              
 Command ===>                                                                   
                                                                    More:     + 
 Data Set Name . . . . : IMSTESTL.IMS01.SPOOL1                                  
                                                                                
 General Data                           Current Allocation                      
  Management class . . : **None**        Allocated tracks  . : 2                
  Storage class  . . . : **None**        Allocated extents . : 1                
   Volume serial . . . : USER02                                                 
   Device type . . . . : 3390                                                   
  Data class . . . . . : **None**                                               
   Organization  . . . : PS             Current Utilization                     
   Record format . . . : VB              Used tracks . . . . : 1                
   Record length . . . : 3180            Used extents  . . . : 1                
   Block size  . . . . : 3190                                                   
   1st extent tracks . : 2                                                      
   Secondary tracks  . : 0              Dates                                   
   Data set name type  :                 Creation date . . . : 2013/11/27       
                                         Referenced date . . : 2023/04/08       
                                         Expiration date . . : ***None***       
   SMS Compressible  . : NO                                                     
                                                                                
  F1=Help      F2=Split     F3=Exit      F7=Backward  F8=Forward   F9=Swap      
 F12=Cancel                                                                     
``` 

New data set `USER.PRIV.PSVB`
```
                              Data Set Information                              
 Command ===>                                                                   
                                                                    More:     + 
 Data Set Name . . . . : USER.PRIV.PSVB                                         
                                                                                
 General Data                           Current Allocation                      
  Management class . . : **None**        Allocated tracks  . : 85               
  Storage class  . . . : **None**        Allocated extents . : 1                
   Volume serial . . . : SCR03                                                  
   Device type . . . . : 3390                                                   
  Data class . . . . . : **None**                                               
   Organization  . . . : PS             Current Utilization                     
   Record format . . . : VB              Used tracks . . . . : 1                
   Record length . . . : 3180            Used extents  . . . : 1                
   Block size  . . . . : 3190                                                   
   1st extent tracks . : 85                                                     
   Secondary tracks  . : 51             Dates                                   
   Data set name type  :                 Creation date . . . : 2023/04/08       
                                         Referenced date . . : 2023/04/08       
                                         Expiration date . . : ***None***       
   SMS Compressible  . : NO                                                     
                                                                                
  F1=Help      F2=Split     F3=Exit      F7=Backward  F8=Forward   F9=Swap      
 F12=Cancel                                                                     
```

**Comment by ketankelkar** [comment]
image

File: plugins/modules/zos_volume_init.py [hunk 19]
```diff
@@ -0,0 +1,249 @@
...
+# limitations under the License.
+
+from __future__ import (absolute_import, division, print_function)
+
+__metaclass__ = type
+
+DOCUMENTATION = r"""
+module: zos_volume_init
+short_description: Initialize volumes or minidisks.
+description:
```
Comment (ddimatos) [inline plugins/modules/zos_volume_init.py:22]: I changed the order up a bit and added some more content around the ICKDFS INIT command with regard to security classes and APF authorization.

I purposely did not make this about RACF and uses the word security classes to allow for a broarder application to  other security tools like TOPSECRET and ACF2. 

Thus consider updating to this:
```
- Initialize a volume or minidisk on z/OS.
- I(zos_volume_init) will create the volume label and entry into the volume
  table of contents (VTOC).
- Volumes are used for storing data and executable programs.
- A minidisk is a portion of a disk that is linked to your virtual machine.
- A VTOC lists the data sets that reside on a volume, their location, size, and
  other attributes.
- I(zos_volume_init) uses the ICKDSF command INIT to initialize a volume. In some
  cases the command could be protected by facility class `STGADMIN.ICK.INIT`.
  Protection occurs when the class is active, and the class profile is defined.
  Ensure the user executing the Ansible task is permitted to execute
  ICKDFSF command INIT, otherwise, any user can use the command.
- ICKDSF is an Authorized Program Facility (APF) program on z/OS,
  I(zos_volume_init) will run in authorized mode but if the program ICKDSF is
  not APF authorized, the task will end.
```

File: plugins/modules/zos_volume_init.py [hunk 20]
```diff
@@ -0,0 +1,249 @@
...
+    required: false
+    default: true
+  verify_volume_empty:
+    description:
+      - Verify that no data sets other than the volume table of contents (VTOC) index or the VSAM Volume Data Set(VVDS) exist on the target volume.
+      - Beware that z/OS system defaults can override ICKDSF parameters.
+    required: false
+    type: bool
+    default: true
+"""
```
Comment (ddimatos) [inline plugins/modules/zos_volume_init.py:103]: Might want to now add a `seealso` section if we mention the zos_backup_restore module.
```
seealso:
- module: zos_backup_restore
```

File: plugins/modules/zos_volume_init.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (ddimatos) [reply plugins/modules/zos_volume_init.py:66]: I think we are good on this, we discussed passing the responsibility down to ICKDSF for validation of special charts and you can escape the special chars with `` or **

File: plugins/modules/zos_volume_init.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ddimatos) [reply plugins/modules/zos_volume_init.py:68]: Waiting on ICKDSF team to advise.

File: plugins/modules/zos_volume_init.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (ketankelkar) [reply plugins/modules/zos_volume_init.py:165]: change to `when ICKDSF is invoked`?

File: plugins/modules/zos_volume_init.py [hunk 21]
```diff
@@ -0,0 +1,249 @@
...
+        "   nods noindex",
+      ]
+msg:
+  description: Failure message returned by module.
+  returned: failure
+  type: str
+  sample: "'Index' cannot be False for SMS managed volumes."
+rc:
+  description:
+    - Return code from ICKDSF init mvs command.
```
Comment (ketankelkar) [inline plugins/modules/zos_volume_init.py:163]: Return code from ICKDSF init ~mvs~ command.

File: plugins/modules/zos_volume_init.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (ddimatos) [reply plugins/modules/zos_volume_init.py:165]: Looks good !!!

File: plugins/modules/zos_volume_init.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (ddimatos) [reply plugins/modules/zos_volume_init.py:169]: Lets go with what you have suggested above `when ICKDSF program is run`

File: plugins/modules/zos_volume_init.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ketankelkar) [reply plugins/modules/zos_volume_init.py:68]: the approach seems to be "if z/OS will allow it, ICKDSF will allow it", decision was to get rid of the "note" line:
~`Characters are not validated so check with operating system guide for valid alphanumeric characters.`~

File: plugins/modules/zos_volume_init.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (ketankelkar) [reply plugins/modules/zos_volume_init.py:66]: When you generated docs, using a pair of backticks \`_char_\` or even pairs/triples of backticks \`\`_char_\`\` seemed to work fine, but when i generated the docs via antsibull (taking the process from the ansible site [here](https://docs.ansible.com/ansible/latest/dev_guide/developing_collections_documenting.html#build-collection-docsite)), i get backticks in my doc

```- Expects 1-6 alphanumeric, national ($, `#`, @) or special characters.```
image

while no backticks (or asterisks) seems to treat the '#' as a comment and disappears everything beyond.

**Comment by ketankelkar** [comment]
during a discussion earlier with @ddimatos, a new issue was uncovered that zos_volume_init module does not facilitate setting `tmphlq` for the underlying call to mvscmdauth (invoked via zoautil_py Python API's mvscmd.execute_authorized).

**Comment by ketankelkar** [comment]
pipeline still good

image

File: plugins/modules/zos_volume_init.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (ddimatos) [reply plugins/modules/zos_volume_init.py]: I am ok with what you have now in the module, they are described well enough, but to answer your question, i don't think the mixed types is an issue because each sub parameter will have a:
type: str | bool | int etc
default: 
etc

So each sub type is really just another type, ill leave it up to you , feel free to resolve this.

File: plugins/modules/zos_volume_init.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ddimatos) [reply plugins/modules/zos_volume_init.py]: I guess there is still a risk that a volume would be one of the noted boolean's so we can go with a description of something like:
```
- Note, this option is **not** a boolean, leave it blank in order to skip the verification.
```

File: plugins/modules/zos_volume_init.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (ddimatos) [reply plugins/modules/zos_volume_init.py]: I see the doc was updated, line seem like the line is a bit long but will be caught pep8 if its an issue.

**Review by ddimatos — CHANGES_REQUESTED** [review]

**Comment by ketankelkar** [comment]
image

**Review by ddimatos — APPROVED** [review]
Great work @ketankelkar , thanks for bearing with me on the reviews.

File: tests/functional/modules/test_zos_volume_init_func.py [hunk 22]
```diff
@@ -0,0 +1,376 @@
...
+        assert result.get("changed") is True
+        assert result.get('rc') == 0
+
+        # The display command issued queries a volume called $TEST_VOL_SER. The
+        # expected return values are 'IEE455I UNIT STATUS NO DEVICES WITH
+        # REQUESTED ATTRIBUTES' or a line with several attributes including unit
+        # address (expected value $TEST_VOL_ADDR) and volume serial (expected
+        # value $TEST_VOL_SER). If those two match, then the 'volid' parameter
+        # is correctly assigned to the 'address' parameter.
+        assert TEST_VOL_SER in display_cmd_output
```
Comment (richp405) [inline tests/functional/modules/test_zos_volume_init_func.py:137]: This assertion concerns me: it is running a test on the same data in each loop, since it is assigned outside of the loop, and no indexing is involved.  I think this means 1 successful match suffices for all results.contacted, which I don't think is the goal... shouldn't it check each result contacted separately, or should it only be called once outside the loop?

**Review by richp405 — APPROVED** [review]
That change to the test is cleaner.  Thanks for that.

**Comment by ketankelkar** [comment]
cicd with new tests looks good!
image
